[Ready for Review] - Make Cairo an optional dependency

pkittenis opened the pull request:
Ready for review.

Changes:
* Cairocffi is *not* pulled in as dependency on Windows _and_ PyPy interpreters via setuptools `extras_requires` environment markers - requires setuptools >= `28.0`
* Without cairo, image/pdf format requests will return 400 error with message that cairo is not installed
* Added tox target for sans-cairo testing of expected 400 error code responses
* Removed cairocffi from tox global dependencies, added to individual env where applicable
* Added setuptools version requirement to tox dependencies
* Added uninstall step for PyPy's `cffi` package which is not compatible with graphite-api. Allows pypy tests to pass sans use of cairo
* Updated readme with note on cairo dependency being optional

To-do:
- [x] Add test target sans cairo
- [x] Add expected `400` error response tests for when cairo is not installed but request is asking for image/pdf formats
- [x] Update readme for `[cairo]` extras parameter on Windows

Comments, suggestions, feedback most welcome.

Resolves #149

brutasse commented on graphite_api/render/glyph.py:
```diff
@@ -12,7 +12,12 @@
-import cairocffi as cairo
+try:
+    import cairocffi as cairo
+except (NameError, ImportError, AttributeError):
```
`except ImportError:` should be enough. What triggered NameError / AttributeError?

pkittenis replied:
`AttributeError` is thrown in cases, like in certain pypy versions, where cairocffi is installed but unusable.

For example on my dev system with pypy 2.2.1:
```
python --version
Python 2.7.3 (2.2.1+dfsg-1ubuntu0.3, Sep 30 2015, 15:18:40)
[PyPy 2.2.1 with GCC 4.8.4]

python -c 'import cairocffi'
Traceback (most recent call last):
  File "app_main.py", line 72, in run_toplevel
  File "app_main.py", line 578, in run_it
  File "<string>", line 1, in <module>
  File "/home/panos/graphite-api/pypyenv/site-packages/cairocffi/__init__.py", line 20, in <module>
    from ._ffi import ffi
  File "/home/panos/graphite-api/pypyenv/site-packages/cairocffi/_ffi.py", line 4, in <module>
    ffi = _cffi_backend.FFI('cairocffi._ffi',
AttributeError: 'module' object has no attribute 'FFI'
```

`NameError` I think is not needed anymore, was using it in an older revision.

brutasse commented on graphite_api/render/glyph.py:
```diff
@@ -12,7 +12,12 @@
+try:
+    import cairocffi as cairo
+except (NameError, ImportError, AttributeError):
+    CAIRO_DISABLED = True
```
You could do `cairo = None` here instead of adding another symbol. Then checks for `CAIRO_DISABLED` would become `if cairo is None`.

pkittenis replied:
Makes sense.

brutasse commented on graphite_api/app.py:
```diff
@@ -553,6 +560,8 @@ def tree_json(nodes, base_path, wildcards=False):
 
 def doImageRender(graphClass, graphOptions):
+    if CAIRO_DISABLED:
+        return
```
An error response is returned earlier L478-481. With an additional check in the `if request_options['graphType'] == 'pie':` branch, you could return an error message there as well and I think this check would not be needed.

brutasse commented:
Actually maybe with something like:

https://wheel.readthedocs.io/en/latest/#defining-conditional-dependencies

```python
extras_require={
    ...
    ':sys_platform!="win32"': ['cairocffi'],
    'cairo': ['cairocffi'],
}
```

It **seems** on non-windows platforms this would pull cairocffi by default and you could `pip install graphite-api[cairo]` on Windows if you really wanted cairocffi.

Thoughts?

pkittenis commented:
Would still be nice to be able to install graphite-api sans cairo on non-Windows platforms.

My use case is not Windows after all but pypy which, at least on some versions, has issues with using, but not installing, cairo.

Also useful for others that just want json/raw format rendering and want to avoid installing all the cairo dependencies, though I guess this could be accomplished by installing as normal then manually uninstalling cairo.

To keep backwards compatibility I guess the least worst option would be leave cairo in requirements and update readme that cairo can be uninstalled. That kind of 'optional requirement' is a bit of a misnomer though :) Thoughts?

brutasse commented on graphite_api/app.py:
```diff
@@ -578,6 +589,8 @@ def tree_json(nodes, base_path, wildcards=False):
 
 
 def doImageRender(graphClass, graphOptions):
+    if not cairo:
```
This shouldn't be silent. What about raising a custom exception here? There are guards in the code so we should never end up here, but for the future it would be nice to fail loudly.

pkittenis replied:
Was thinking it would be better not to fail actually. Since we do want it to fail loudly, will remove that check and let the underlying exception get raised.

brutasse commented:
If cairo installs with PyPy but doesn't load, we can keep it in the requirements, no? Or is the exception raised later than when importing?

If so, https://github.com/brutasse/graphite-api/pull/198#issuecomment-253747594 should still work.

If not, we can do a cairo render check at boot time and disable cairo if it fails…

pkittenis commented:
Exception is raised at import and is handled, if cairo is really optional, though, should it be in `requirements`?

pkittenis commented:
This is now ready for review. Have added a `no-cairo` tox target that tests without cairo installed and updated tests with expected responses and error codes/messages for when cario is not available.

Readme also updated with a note on how cairo is now optional and how to re-enable on Windows.

pkittenis commented:
Pypy/Pyparsing failures do not seem to come from these changes, also get them on current master.

pkittenis commented:
Ideally cairo should be disabled by default on PyPy as well, many issues there also. Suggestions welcome on how to do that.

pkittenis commented:
Have added a change to not install cairo on PyPy as well. PyPy has its own cffi implementation which could be used - the pip cairocffi package is not compatible. 

However, pypy's cffi implementation (included by default on recent versions of pypy) is not compatible with graphite-api and is causing pypy tests to fail. Have changed pypy tox env to explicitly uninstall it.

Had to add a minimum setuptools version in tox dependencies for this to work (on travis it is already met, may not be the case locally).

Environment markers requires at least setuptools version `28.0`.

pkittenis commented:
Rebased on latest master.

pkittenis commented:
Anything else still needed on this PR?
